text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
Some possible points in favor, weak and not backed by citations but maybe useful for starting this discussion:

Async I/O avoids context switches, which usually mean a call into the kernel and shuffling registers to sleep, and then another one along the same lines to wake back up.  You're also going to be thrashing the CPU cache by completely switching out the stack hundreds of thousands of times a second or more.  None of your data for anything is local and even really, really small ones end up taking at least a page.  I could see both of these alone being detrimental under incredibly heavy Google-scale workloads.

Every OS vendor I know of has felt the need to go above and beyond Berkeley sockets in terms of providing async I/O functions.  The effort put into this is very far from trivial, and OS vendors don't do things just because.  This is either a historical quirk or because OS vendors think you can do better than they can with application-specific scheduling.

Google decided to go with green threads (which are essentially stackful coroutines) for Go, which is a recent-ish language.  Just relying on the OS would have been an equally valid implementation, what they did clearly took significant effort, and yet they did it anyway.

There's an issue with combining tens of thousands of I/O threads and a few compute threads that's hard to get right: either your compute threads get high priority and slow down I/O, or your compute threads get low priority and get preempted like hell.  You can deal with this by playing around with thread affinities, but you'll have to set priority and affinity on every new thread.  This has overhead, but even deciding what the affinity needs to be in the first place isn't trivial.  You can use a thread pool that's allowed to grow as large as needed to deal with the overhead of setting thread-specific settings, but you'll need algorithms on par with just having used async/await in the first place to deal with scheduling and despawning unneeded threads.  And don't forget about those rare times when something beyond your control uses thread-local storage, so now you have to expose CPU pinning to the user.

Or maybe there's a clever multithreaded solution that allows you to deal with all of this by mostly sweeping it under the rug that I don't know about.  My area of expertise isn't this sort of thing, so I wouldn't be surprised to find out I missed developments.  Nonetheless async/await lets you efficiently pretend that you're writing synchronous code and get advantages, and I don't see how using threads lets you get away with this at all.